File: NSTU_Modelirovanie_Processov_and_sysrem/lab/lab3/main_form_dispace.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <random>
#include <ctime>

using namespace std;

const int NUM_READERS = 3; // Количество процессов чтения
const int BUFFER_SIZE = 10; // Размер буфера записи
const string FILENAME = "data.txt";

struct Reader {
    vector<char> buffer;
    int read_count = 0;
};

void initializeFile() {
    ofstream file(FILENAME, ios::trunc);
    if (!file) {
        cerr << "Ошибка открытия файла!" << endl;
        exit(1);
    }
    file << "abcdefghijklmnopqrstuvwxyz"; // Записываем тестовые данные
    file.close();
}

int weightedRandomSelection(const vector<double>& probabilities) {
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<double> dist(0.0, 1.0);
    double r = dist(gen);
    double sum = 0;
    for (int i = 0; i < probabilities.size(); i++) {
        sum += probabilities[i];
        if (r < sum) return i;
    }
    return probabilities.size() - 1;
}

int main() {
    srand(time(0));
    initializeFile();

    Reader readers[NUM_READERS];
    vector<char> writeBuffer;
    ifstream inFile(FILENAME);
    ofstream outFile;
    bool writing = false;

    vector<double> probabilities(NUM_READERS + 1, 1.0 / (NUM_READERS + 1));

    int totalWritten = 0;
    while (true) {
        int selectedProcess = weightedRandomSelection(probabilities);

        if (selectedProcess < NUM_READERS) { // Читающий процесс
            if (!writing) {
                char ch;
                if (!(inFile >> ch)) {
                    inFile.clear();
                    inFile.seekg(0, ios::beg);
                    inFile >> ch;
                }
                readers[selectedProcess].buffer.push_back(ch);
                readers[selectedProcess].read_count++;
            }
        } else { // Процесс записи
            if (!writing) {
                writing = true;
                outFile.open(FILENAME, ios::trunc);
                if (!outFile) {
                    cerr << "Ошибка открытия файла для записи!" << endl;
                    exit(1);
                }
            }
            if (!writeBuffer.empty()) {
                outFile << writeBuffer.back();
                writeBuffer.pop_back();
                totalWritten++;
            } else {
                outFile.close();
                break;
            }
        }
    }

    cout << "Результаты чтения:" << endl;
    for (int i = 0; i < NUM_READERS; i++) {
        cout << "Читатель " << i + 1 << " считал " << readers[i].read_count << " символов." << endl;
    }
    return 0;
}

File: NSTU_Modelirovanie_Processov_and_sysrem/lab/lab3/main_form_dispace.py
import random
import time

FILENAME = "data.txt"
NUM_READERS = 3  # Количество процессов чтения
BUFFER_SIZE = 10  # Размер буфера записи

class Reader:
    def __init__(self):
        self.buffer = []
        self.read_count = 0

def initialize_file():
    with open(FILENAME, "w") as f:
        f.write("abcdefghijklmnopqrstuvwxyz")

def weighted_random_selection(probabilities):
    r = random.uniform(0, 1)
    cumulative = 0
    for i, p in enumerate(probabilities):
        cumulative += p
        if r < cumulative:
            return i
    return len(probabilities) - 1

def main():
    initialize_file()
    
    readers = [Reader() for _ in range(NUM_READERS)]
    write_buffer = []
    writing = False
    
    probabilities = [1.0 / (NUM_READERS + 1)] * (NUM_READERS + 1)
    total_written = 0
    
    with open(FILENAME, "r+") as f:
        while True:
            selected_process = weighted_random_selection(probabilities)
            
            if selected_process < NUM_READERS:  # Читающий процесс
                if not writing:
                    ch = f.read(1)
                    if not ch:  # Если дошли до конца файла, читаем с начала
                        f.seek(0)
                        ch = f.read(1)
                    readers[selected_process].buffer.append(ch)
                    readers[selected_process].read_count += 1
            else:  # Процесс записи
                if not writing:
                    writing = True
                    f.seek(0)
                    f.truncate()
                
                if write_buffer:
                    f.write(write_buffer.pop())
                    total_written += 1
                else:
                    break
            
            time.sleep(0.01)  # Небольшая задержка для имитации процессов
    
    print("Результаты чтения:")
    for i, reader in enumerate(readers):
        print(f"Читатель {i + 1} считал {reader.read_count} символов.")

if __name__ == "__main__":
    main()
